source/crud/bulk-write.md

```diff
@@ -0,0 +1,814 @@
+
+- Expose the `hasVerboseResults` field in `BulkWriteResult` as defined above. Document that
+  `insertResults`, `updateResults`, and `deleteResults` will be undefined when `hasVerboseResults`
+  is false. Raise an error if a user tries to access one of these fields when `hasVerboseResults`
+  is false.
+- Implement the `insertResults`, `updateResults`, and `deleteResults` fields as optional types and
+  document that they will be unset when `verboseResults` is false.
+- Introduce separate `SummaryBulkWriteResult` and `VerboseBulkWriteResult` types.
+  `VerboseBulkWriteResult` MUST have all of the required fields defined on `BulkWriteResult` above.
+  `SummaryBulkWriteResult` MUST have all of the required fields defined on `BulkWriteResult` above
+  except `insertResults`, `updateResults`, and `deleteResults`.
+
+#### Individual results
+
+The `InsertOneResult`, `UpdateResult`, and `DeleteResult` classes are the same as or similar to
+types of the same name defined in the [CRUD specification](crud.md). Drivers MUST redefine these
+classes if their existing result classes deviate from the definitions in this specification (e.g.
+if they contain acknowledgement information, which is not applicable for individual bulk write
+operations). Drivers MAY reuse their existing types for these classes if they match the ones
+defined here exactly.
+
+### Exception
+
+```typescript
+class BulkWriteException {
+    /**
+     * A top-level error that occurred when attempting to communicate with the server or execute
+     * the bulk write. This value may not be populated if the exception was thrown due to errors
+     * occurring on individual writes.
+     */
+    error: Optional<Error>;
+
+    /**
+     * Write concern errors that occurred while executing the bulk write. This list may have
+     * multiple items if more than one server command was required to execute the bulk write.
+     */
+    writeConcernErrors: WriteConcernError[];
+
+    /**
+     * Errors that occurred during the execution of individual write operations. This map will
+     * contain at most one entry if the bulk write was ordered.
+     */
+    writeErrors: Map<Int64, WriteError>;
+
+    /**
+     * The results of any successful operations that were performed before the error was
+     * encountered.
+     */
+    partialResult: Optional<BulkWriteResult>;
+}
+```
+
+## Building a `bulkWrite` Command
+
+The `bulkWrite` server command has the following format:
+
+```json
+{
+    "bulkWrite": 1,
+    "ops": <Array>,
+    "nsInfo": <Array>,
+    "errorsOnly": Optional<Boolean>,
+    "ordered": Optional<Boolean>,
+    "bypassDocumentValidation": Optional<Boolean>,
+    "comment": Optional<BSON value>,
+    "let": Optional<Document>,
+    ...additional operation-agnostic fields
+}
+```
+
+Drivers MUST use document sequences ([`OP_MSG`](../message/OP_MSG.rst) payload type 1) for the
+`ops` and `nsInfo` fields. Drivers MUST NOT use document sequences when auto-encryption is enabled.
```

Comment:
`nsInfo` could exceed `maxBsonObjectSize` (~16MB) if

1. a user is doing `maxWriteBatchSize` operations,
2. each operation is on a different namespace, and
3. each namespace is close to the maximum length defined [here](https://www.mongodb.com/docs/manual/reference/limits/#mongodb-limit-Restriction-on-Collection-Names).

Just the namespaces would be 25MB in this scenario, and then there would be additional bytes to wrap each namespace in a document and then put those documents into an array.

Large `nsInfo` payloads could cause earlier batch splitting when they're used within the command document rather than as a document sequence. Batch splitting logic may also become more complex if adding one model to a bulk write `OP_MSG` requires size checks for both the command document and the overall message.

How much of a lift would it be to support multiple document sequences in the C driver? I'm inclined to leave in this requirement and expect drivers to update their `OP_MSG` implementations to support multiple document sequences, which is in line with the `OP_MSG` spec. FWIW, I made this change in the Rust driver https://github.com/mongodb/mongo-rust-driver/pull/1009/and it wasn't too much work.

Reply:
> How much of a lift would it be to support multiple document sequences in the C driver?

It was not significant. Change has been applied to the C implementation.

Retracting suggestion.

Reply:
Sounds good, I'll add a Q&A entry for this.